src/lib/optimizer/strategy/join_to_semi_join_rule.cpp

```diff
@@ -72,20 +74,30 @@ void JoinToSemiJoinRule::_apply_to_plan_without_subqueries(const std::shared_ptr
       }
 
       // Determine which node to use for Semi-Join-filtering and check for the required uniqueness guarantees.
-      if (*join_node->prunable_input_side() == LQPInputSide::Left &&
-          join_node->left_input()->has_matching_ucc(equals_predicate_expressions_left)) {
-        join_node->join_mode = JoinMode::Semi;
-        const auto temp = join_node->left_input();
-        join_node->set_left_input(join_node->right_input());
-        join_node->set_right_input(temp);
-      } else if (*join_node->prunable_input_side() == LQPInputSide::Right &&
-                 join_node->right_input()->has_matching_ucc(equals_predicate_expressions_right)) {
-        join_node->join_mode = JoinMode::Semi;
+      if (*join_node->prunable_input_side() == LQPInputSide::Left) {
+        const auto matching_ucc = join_node->left_input()->get_matching_ucc(equals_predicate_expressions_left);
+        if (matching_ucc.has_value()) {
+          rule_was_applied_using_non_permanent_ucc = !matching_ucc->is_permanent();
+
+          join_node->join_mode = JoinMode::Semi;
+          const auto temp = join_node->left_input();
+          join_node->set_left_input(join_node->right_input());
+          join_node->set_right_input(temp);
+        }
+      } else if (*join_node->prunable_input_side() == LQPInputSide::Right) {
+        const auto matching_ucc = join_node->right_input()->get_matching_ucc(equals_predicate_expressions_right);
+        if (matching_ucc.has_value()) {
+          rule_was_applied_using_non_permanent_ucc = !matching_ucc->is_permanent();
+
+          join_node->join_mode = JoinMode::Semi;
+        }
       }
     }
 
     return LQPVisitation::VisitInputs;
   });
+
+  return rule_was_applied_using_non_permanent_ucc ? IsCacheable::No : IsCacheable::Yes;
```

Comment:
pick here: if you implemented operator `|=` for `IsCacheable` taking a bool as second param, you could simplify the rules a bit. Not sure if maybe the data dependencies should also just have a member of type `IsCacheable` rather than the `permanent` bool.